default/python/AI/combat_simulator.py

```diff
@@ -0,0 +1,412 @@
+                        self.faction_B.add_ship(design_name=this_line)
+
+
+    def add_ship_for_A(self, design_id=None, design_name=None):
+        self.faction_A.add_ship(design_id, design_name)
+
+    def add_planet_for_A(self, defense, shields):
+        self.faction_A.add_planet(defense, shields)
+
+    def add_ship_for_B(self, design_id=None, design_name=None):
+        self.faction_B.add_ship(design_id, design_name)
+
+    def add_planet_for_B(self, defense, shields):
+        self.faction_B.add_planet(defense, shields)
+
+    def run(self, iterations):
+        self.__init__()
+        self.parse_shiplist()
+
+        WIN_A = -1
+        WIN_B = 1
+        DRAW = 0
+
+        results = []
+        lost_pp_A = []
+        lost_pp_B = []
+
+        for i in xrange(iterations):
+            self.faction_A.reset()
+            self.faction_B.reset()
+
+            original_number_of_ships_A = self.faction_A.remaining_ships()
+            original_pp_A = self.faction_A.remaining_pp()
+
+            original_number_of_ships_B = self.faction_B.remaining_ships()
+            original_pp_B = self.faction_B.remaining_pp()
+            if i == 0:
+                chat_human("Number of ships A: %d (cost: %.1f PP)" % (original_number_of_ships_A, original_pp_A))
+                chat_human("Number of ships B: %d (cost: %.1f PP)" % (original_number_of_ships_B, original_pp_B))
+
+            for combat_round in xrange(3):
+                self.faction_A.shoot(self.faction_B.current_objects)
+                self.faction_B.shoot(self.faction_A.current_objects)
+                self.faction_A.launch_fighters()
+                self.faction_B.launch_fighters()
+                self.faction_A.update()
+                self.faction_B.update()
+                if not self.faction_A.is_alive() or not self.faction_B.is_alive():
+                    break
+
+            remaining_pp_A = self.faction_A.remaining_pp()
+            lost_pp_A.append(original_pp_A - remaining_pp_A)
+
+            remaining_pp_B = self.faction_B.remaining_pp()
+            lost_pp_B.append(original_pp_B - remaining_pp_B)
+
+            if original_number_of_ships_A:
+                win_a = self.faction_A.remaining_ships() > 0
+            else:
+                win_a = self.faction_A.remaining_planets() > 0
+            if original_number_of_ships_B:
+                win_b = self.faction_B.remaining_ships() > 0
+            else:
+                win_b = self.faction_B.remaining_planets() > 0
+
+            if win_a and win_b:
+                results.append(DRAW)
+            elif win_a:
+                results.append(WIN_A)
+            elif win_b:
+                results.append(WIN_B)
+            else:
+                results.append(DRAW)
+
+        mean_losses_A = mean(lost_pp_A)
+        std_losses_A = std(lost_pp_A)
+        mean_losses_B = mean(lost_pp_B)
+        std_losses_B = std(lost_pp_B)
+
+        draws = len([r for r in results if r == DRAW])
```

Comment:
`len(r for r in results if r == DRAW)`  # generator expression
`sum(r == DRAW for r in results)`